Libc/sys/sem_open.2
.\"	$Darwin$
.\"
.\" Copyright (c) 2000-2002 Apple Computer, Inc. All rights reserved.
.\"
.\" @APPLE_LICENSE_HEADER_START@
.\" 
.\" The contents of this file constitute Original Code as defined in and
.\" are subject to the Apple Public Source License Version 1.1 (the
.\" "License").  You may not use this file except in compliance with the
.\" License.  Please obtain a copy of the License at
.\" http://www.apple.com/publicsource and read it before using this file.
.\" 
.\" This Original Code and all software distributed under the License are
.\" distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, EITHER
.\" EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
.\" INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT.  Please see the
.\" License for the specific language governing rights and limitations
.\" under the License.
.\" 
.\" @APPLE_LICENSE_HEADER_END@
.\"
.Dd June 8, 2000
.Dt SEM_OPEN 2
.Os Darwin
.Sh NAME
.Nm sem_open
.Nd initialize and open a named semaphore
.Sh SYNOPSIS
.Fd #include <semaphore.h>
.Ft sem_t *
.Fn sem_open "const char *name" "int flags"
.Ft sem_t *
.Fn sem_open "const char *name" "int flags" "mode_t mode" "unsigned int value"
.Sh DESCRIPTION
The named semaphore named
.Fa name
is initialized and opened as specified by the argument
.Fa flags
and a semaphore descriptor is returned to the calling process.
.Pp
The flags specified are formed by
.Em or Ns 'ing
the following values:
.Pp
.Bd -literal -offset indent -compact
O_CREAT		create the semaphore if it does not exist
O_EXCL		error if create and semaphore exists
.Ed
.Pp
If
.Dv O_CREATE
if specified,
.Fn sem_open
requires an additional two arguments.
.Fa mode
specifies the permissions for the semaphore as described in
.Xr chmod 2
and modified by the process' umask value (see
.Xr umask 2 ) .
The semaphore is created with an initial
.Fa value ,
which must be less than or equal to
.Dv SEM_VALUE_MAX .
.Pp
If
.Dv O_EXCL
is specified and the semaphore exists,
.Fn sem_open
fails.  The check for the existence of the semaphore and the creation
of the semaphore are atomic with respect to all processes calling
.Fn sem_open
with
.Dv O_CREAT
and
.Dv O_EXCL
set.
.Pp
When a new semaphore is created, it is given the user ID and group ID
which coorespond to the effective user and group IDs of the calling
process. There is no visible entry in the file system for the created
object in this implementation.
.Pp
The returned semaphore descriptor is available to the calling process
until it is closed with
.Fn sem_close ,
or until the caller exits or execs.
.Pp
If a process makes repeated calls to
.Fn sem_open ,
with the same
.Fa name
argument, the same descriptor is returned for each successful call,
unless
.Fn sem_unlink
has been called on the semaphore in the interim.
.Pp
If
.Fn sem_open
fails for any reason, it will return a value of
.Dv SEM_FAILED
and sets
.Va errno .
On success, it returns a semaphore descriptor.
.Sh ERRORS
The named semaphore is opened unless:
.Bl -tag -width Er
.It Bq Er EACCES
The required permissions (for reading and/or writing)
are denied for the given flags; or
.Dv O_CREAT
is specified, the object does not exist, and permission to
create the semaphore is denied.
.It Bq Er EEXIST
.Dv O_CREAT
and
.Dv O_EXCL
were specified and the semaphore exists.
.It Bq Er EINTR
The
.Fn sem_open
operation was interrupted by a signal.
.It Bq Er EINVAL
The
.Fn shm_open
operation is not supported; or
.Dv O_CREAT
is specified and
.Fa value
exceeds
.Dv SEM_VALUE_MAX .
.It Bq Er EMFILE
The process has already reached its limit for semaphores or file
descriptors in use.
.It Bq Er ENAMETOOLONG
.Fa name
exceeded
.Dv SEM_NAME_LEN
characters.
.It Bq Er ENFILE
Too many semaphores or file descriptors are open on the system.
.It Bq Er ENOENT
.Dv O_CREAT
is not set and the named semaphore does not exist.
.It Bq Er ENOSPC
.Dv O_CREAT
is specified, the file does not exist, and there is insufficient
space available to create the semaphore.
.El
.Sh SEE ALSO
.Xr semctl 2 ,
.Xr semget 2 ,
.Xr semop 2 ,
.Xr sem_close 2 ,
.Xr sem_post 2 ,
.Xr sem_trywait 2 ,
.Xr sem_unlink 2 ,
.Xr sem_wait 2 ,
.Xr umask 2
.Sh HISTORY
.Fn sem_open
is specified in the POSIX Realtime Extension (1003.1b-1993/1003.1i-1995).


X11/ChangeLog
2003-09-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: revert fix for
	3363492 and add another check for RootlessKnowsWindowNumber ()
	to fix the original bug (3420745)

2003-09-10  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c (sigchld_handler):
	loop until no more zombies exist (3414442)

2003-08-08  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m,
	xc/programs/Xserver/hw/apple/X11Application.h: move calling
	HideMenuBar and ShowMenuBar onto the main UI thread (3359753)

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	few more MT changes, be even more careful about when contexts
	are updated

2003-08-07  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	Large set of changes to make this code handle multiple threads
	sanely. Changes include adding a mutex for each screen, and
	only ever touching CGL contexts when we know they're bound to
	the current thread or no thread. This involves patching gli
	dispatch table to cause glViewport to call back into DRI code
	so that contexts can be updated on their own thread. May be
	occasions where glXSwapBuffers fails if no suitable context can
	be found to issue the swap on (3340377) 

	* xc/programs/Xserver/hw/apple/quartz.c: quartzServerVisible
	variable needs to start out as false (3366845)

2003-08-04  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: move the call
	to RootlessKnowsWindowNumber () into the event handler so that
	we don't activate X on the broken menu-mouse-down event
	(3363492)

	* RELEASE-NOTES: updated to reflect that X11 is now part of the
	OS and no longer a preview release (3361286)

2003-07-29  John Harper  <jsh@apple.com>

	* xc/lib/GL/glx/glxext.c (glXMakeCurrent): unlock display after
	sending GLXMakeCurrent request when changing from indirect to
	direct context (3351489, sent upstream via Torrey)

2003-07-25  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	when in fullscreen mode, don't pass ApplicationDidReactivate
	events through to AppKit (3341360)

2003-07-23  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.c (XAppleDRIGetIndirectContext):
	make the context we allocate 1k bigger than we think it needs
	to be, in case GL changes

2003-07-22  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.c (XAppleDRIGetIndirectContext):
	allocate the CGLContext used for dispatch indirect rendering
	ourselves, avoids ever needing to contact the window server
	(3126715)

2003-07-18  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/dri-surface.c: fixed use of
	pixmap private index

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzSetFullscreen):
	use xp_disable_hot_keys instead of CG spi (3221162)

	* xc/programs/Xserver/hw/apple/rootless-screen.c,
	xc/programs/Xserver/hw/apple/rootless-common.h,
	xc/programs/Xserver/hw/apple/rootless-common.c: use
	RootlessFinishedDrawing () to call new function
	RootlessMayRedisplay (). After a certain amount of time since
	the last update it will cause a flush (3270678)

	* xc/programs/Xserver/hw/apple/rootless-screen.c: clean-up

2003-07-17  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless-window.c
	(RootlessPaintWindowBackground): prevent trying to draw in
	surfaces here

	* xc/programs/Xserver/hw/apple/rootless-gc.c: call
	RootlessFinishedDrawing at the end of each rendering primitive

	* xc/programs/Xserver/hw/apple/rootless-common.h,
	xc/programs/Xserver/hw/apple/rootless-common.c: use the
	DRIStartDrawing functions etc to test if the drawable is a
	surface.
	(RootlessFinishedDrawing): new function, should be called after
	drawing has finished

	* xc/programs/Xserver/hw/apple/dristruct.h,
	xc/programs/Xserver/hw/apple/dri-surface.h,
	xc/programs/Xserver/hw/apple/dri-surface.c (DRIStartDrawing,
	DRIStopDrawing, DRIDamageRegion, DRISynchronizeDrawable,
	DRISynchronize): functions to support Xlib rendering on
	accelerated surfaces. Only supported on 10.3, and they won't do
	anything by default

	* xc/programs/Xserver/hw/apple/dri-surface.c
	(DRICreateSurface): support Pixmap drawables (3122756, 10.3
	only)

	* xc/programs/Xserver/hw/apple/dri-surface.c: removed check for
	cg from SUJagGreen

	* xc/programs/Xserver/GL/apple/indirect.c: support offscreen
	drawables (by removing the code that stopped them being
	supported, dri-surface.c does the real work) (3122756)

	* xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/quartz-server.c,
	xc/programs/Xserver/hw/apple/X11Application.m,
	xc/programs/Xserver/hw/apple/X11Application.h: added xp_options
	default, lets you pass extra init options to Xplugin

	* xc/programs/Xserver/hw/apple/*: fixed gcc 3.3 warnings

2003-07-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/x-list.c,
	xc/programs/Xserver/hw/apple/x-list.h (x_list_pop): new
	function

	* xc/programs/Xserver/hw/apple/x-hash.c: fix gcc 3.3 warning

2003-07-07  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: don't pass
	flags changed events to X11 when Appkit is active (3316333)

2003-07-02  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzRunClient):
	don't quote commands, it stops multi-word commands working

	* xc/programs/Xserver/hw/apple/X11Controller.m: updated quit
	alert text (3313133)

	* xc/programs/Xserver/hw/apple/Xquartz.man: added more defaults

	* xc/programs/Xserver/hw/apple/bundle-main.c (main): detach
	from the controlling terminal when we have one (avoids bash
	suspending the process when we try to launch problems!?)

2003-06-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/applewm-impl.c
	(ProcAppleWMSetWindowLevel): removed the code to call

Libc/i386/sys/sem_open.s
/*
 * Copyright (c) 1999 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * The contents of this file constitute Original Code as defined in and
 * are subject to the Apple Public Source License Version 1.1 (the
 * "License").  You may not use this file except in compliance with the
 * License.  Please obtain a copy of the License at
 * http://www.apple.com/publicsource and read it before using this file.
 * 
 * This Original Code and all software distributed under the License are
 * distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT.  Please see the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
#include "SYS.h"

UNIX_SYSCALL(sem_open, 4)
	ret

